### PR TITLE
Better exception message when using custom migrations

### PR DESCRIPTION
**Pull request by cmelchior**
People not overriding `equals` in their custom Migration class is a common source of errors if people are creating multiple RealmConfiguration instances (which they shouldn't in the first place), but at least this gives better hints to what is wrong.

@realm/java

### CODE REVIEW

**Comment by dalinaum** [comment]
:+1:

File: realm/realm-library/src/androidTest/java/io/realm/RealmConfigurationTests.java [hunk 1]
```diff
@@ -779,4 +779,27 @@ public void initialDataTransactionAssetFile() throws IOException {
...
+    public void detectMissingEqualsInCustomMigration() {
+        RealmConfiguration config1 = configFactory.createConfigurationBuilder().migration(new MigrationWithNoEquals()).build();
+        RealmConfiguration config2 = configFactory.createConfigurationBuilder().migration(new MigrationWithNoEquals()).build();
+
+        Realm realm = Realm.getInstance(config1);
+        try {
+            Realm.getInstance(config2);
+            fail();
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("The most likely cause are that equals() and hashCode() are not overridden"));
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmConfigurationTests.java]: `cause are` -> `cause is` or `causes are`?

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 2]
```diff
@@ -223,6 +223,16 @@ private void validateConfiguration(RealmConfiguration newConfiguration) {
         if (!Arrays.equals(configuration.getEncryptionKey(), newConfiguration.getEncryptionKey())) {
             throw new IllegalArgumentException(DIFFERENT_KEY_MESSAGE);
         } else {
+            // A common problem is that people are forgetting to override `equals` in their custom migration class.
+            // Try to detect this problem specifically so we can throw a better error message.
+            RealmMigration newMigration = newConfiguration.getMigration();
+            RealmMigration oldMigration = configuration.getMigration();
+            if (oldMigration != null && newMigration != null && oldMigration.getClass().equals(newMigration.getClass())) {
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: We should have another condition `!newMigration.equals(oldMigration )`.

**Comment by zaki50** [comment]
:+1: besides small comments.

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 3]
```diff
@@ -223,6 +223,16 @@ private void validateConfiguration(RealmConfiguration newConfiguration) {
...
+                throw new IllegalArgumentException("Configurations cannot be different if used to open the same file. " +
+                        "The most likely cause are that equals() and hashCode() are not overridden in the " +
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: `cause are` -> `cause is`?

**Comment by beeender** [comment]
:+1:

**Comment by kneth** [comment]
:+1:

**Pull request merged**